project/Protobuf.scala

```diff
@@ -1,87 +1,131 @@
+      thunk(proc, log)
+    } catch { case e: Exception =>
+      throw new RuntimeException("error while executing '%s' with args: %s" format(protoc, args.mkString(" ")), e)
+    }
+
+  private def checkProtocVersion(protoc: String, protocVersion: String, log: Logger): Unit = {
+    val res = callProtoc(protoc, Seq("--version"), log, { (p, l) => p !! l })
+    val version = res.split(" ").last.trim
+    if (version != protocVersion) {
+      sys.error("Wrong protoc version! Expected %s but got %s" format (protocVersion, version))
+    }
+  }
+
+  private def generate(protoc: String, srcDir: File, targetDir: File, log: Logger): Unit = {
+    val protoFiles = (srcDir ** "*.proto").get
+    if (srcDir.exists)
+      if (protoFiles.isEmpty)
+        log.info("Skipping empty source directory %s" format srcDir)
+      else {
+        targetDir.mkdirs()
+
+        log.info("Generating %d protobuf files from %s to %s".format(protoFiles.size, srcDir, targetDir))
+        protoFiles.foreach { proto => log.info("Compiling %s" format proto) }
+
+        val exitCode = callProtoc(protoc, Seq("-I" + srcDir.absolutePath, "--java_out=%s" format targetDir.absolutePath) ++
+          protoFiles.map(_.absolutePath), log, { (p, l) => p ! l })
+        if (exitCode != 0)
+          sys.error("protoc returned exit code: %d" format exitCode)
+      }
+  }
+
+  /**
+   * Create a transformed version of all files in a directory, given a predicate and a transform function for each file. From sbt-site
+   */
+  private def transformDirectory(sourceDir: File, targetDir: File, transformable: File => Boolean, transform: (File, File) => Unit, cache: File, log: Logger): File = {
+    val runTransform = FileFunction.cached(CacheStoreFactory(cache), FilesInfo.hash, FilesInfo.exists) { (in: ChangeReport[File], out: ChangeReport[File]) =>
+      val map = Path.rebase(sourceDir, targetDir)
+      if (in.removed.nonEmpty || in.modified.nonEmpty) {
+        log.info("Preprocessing directory %s..." format sourceDir)
+        for (source <- in.removed; target <- map(source)) {
+          IO delete target
+        }
+        val updated = for (source <- in.modified; target <- map(source)) yield {
+          if (source.isFile) {
+            if (transformable(source)) transform(source, target)
+            else IO.copyFile(source, target)
+          }
+          target
+        }
+        log.info("Directory preprocessed: " + targetDir)
+        updated
+      } else Set.empty
+    }
+    val sources = (sourceDir.allPaths).get.toSet
+    runTransform(sources)
+    targetDir
+  }
+
+  /**
+   * Transform a file, line by line.
+   */
+  def transformFile(transform: String => String)(source: File, target: File): Unit = {
+    IO.reader(source) { reader =>
+      IO.writer(target, "", IO.defaultCharset) { writer =>
+        val pw = new PrintWriter(writer)
+        IO.foreachLine(reader) { line => pw.println(transform(line)) }
+      }
+    }
+  }
+
+
+}
```

Comment:
Oh so extracted the feature we used here from sbt-site?

Reply:
Yeah, it became private there, so I took the 'easy' way out and extracted it.

Reply:
👍  good; power of open source ;-)